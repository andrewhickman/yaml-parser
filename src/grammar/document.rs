use crate::{
    char, cursor::Cursor, diag::DiagnosticKind, parser::Buffer, Diagnostic, Receiver, Token,
};

use super::{trivia::comment_lines, try_token_char};

pub(super) fn prefix<'s>(
    cursor: &mut Cursor<'s>,
    receiver: &mut impl Receiver,
    prev_terminated: bool,
) -> Result<(), Diagnostic> {
    try_token_char(
        cursor,
        receiver,
        Token::ByteOrderMark,
        char::BYTE_ORDER_MARK,
    )?;

    comment_lines(cursor, receiver)?;

    while cursor.is_char(char::DIRECTIVE)? {
        if !cursor.is_start_of_line() {
            // todo recover?
            // return Err(Error::new(ErrorKind::DirectiveNotAtStartOfLine, cursor.next_char_span()))
        }
        if !prev_terminated {
            // todo recover?
            // return Err(Error::new(ErrorKind::DirectiveAfterUnterminatedDocument, cursor.next_char_span()))
        }
    }
    // if cursor.is_char(ch)

    todo!()
}

// fn recover_directive
