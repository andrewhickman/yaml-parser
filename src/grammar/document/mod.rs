#[cfg(test)]
mod tests;

use alloc::{
    borrow::{Cow, ToOwned},
    collections::{btree_map, BTreeMap},
};

use crate::{
    char,
    cow::CowBuilder,
    cursor::Cursor,
    diag::{DiagnosticKind, Expected},
    grammar::{recover, tag, token_char},
    Diagnostic, Receiver, Span, Token,
};

use super::{
    trivia::{self, trailing_line},
    try_token_char,
};

#[derive(Clone, Debug, Default)]
#[cfg_attr(test, derive(serde::Serialize), serde(rename_all = "lowercase"))]
pub(crate) struct Document<'s> {
    explicit: bool,
    version: Option<Cow<'s, str>>,
    tags: BTreeMap<Cow<'s, str>, Cow<'s, str>>,
}

pub(super) fn prefix<'s>(
    cursor: &mut Cursor<'s>,
    receiver: &mut (impl Receiver + ?Sized),
    prev_terminated: bool,
) -> Result<(Document<'s>, Span), Diagnostic> {
    try_token_char(
        cursor,
        receiver,
        Token::ByteOrderMark,
        char::BYTE_ORDER_MARK,
    )?;

    trivia::comment_lines(cursor, receiver)?;

    let mut document = Document::default();
    let start = cursor.location();

    let have_directives = cursor.is_char(char::DIRECTIVE)?;
    if have_directives {
        if !prev_terminated {
            receiver.diagnostic(Diagnostic::new(
                DiagnosticKind::DirectiveAfterUnterminatedDocument,
                cursor.next_span(),
            ));
        }

        directives(cursor, receiver, &mut document)?;
    }

    if cursor.is_str("---")? {
        let is_start_of_line = cursor.is_start_of_line();
        let followed_by_whitespace =
            matches!(cursor.peek_nth(3)?, None | Some('\r' | '\n' | '\t' | ' '));

        if (is_start_of_line && followed_by_whitespace) || have_directives || !prev_terminated {
            cursor.eat_str("---")?;
            let span = cursor.token();
            receiver.token(Token::DirectivesEnd, span);

            if !is_start_of_line {
                receiver.diagnostic(Diagnostic::new(
                    DiagnosticKind::DirectivesEndNotAtStartOfLine,
                    span,
                ));
            }

            if followed_by_whitespace {
                trailing_line(cursor, receiver)?;
            } else {
                receiver.diagnostic(Diagnostic::new(
                    DiagnosticKind::DirectivesEndNotFollowedByWhitespace,
                    span,
                ));
            }

            document.explicit = true;
        }
    } else if have_directives {
        receiver.diagnostic(Diagnostic::new(
            DiagnosticKind::MissingDirectivesEndAfterDirective,
            cursor.empty_span(),
        ));
    } else if !prev_terminated {
        receiver.diagnostic(Diagnostic::new(
            DiagnosticKind::MissingDirectivesEndAfterUnterminatedDocument,
            cursor.empty_span(),
        ));
    }

    let span = cursor.span(start);

    trivia::comment_lines(cursor, receiver)?;

    Ok((document, span))
}

pub(super) fn suffix(
    cursor: &mut Cursor,
    receiver: &mut (impl Receiver + ?Sized),
) -> Result<Span, Diagnostic> {
    debug_assert!(cursor.is_token_boundary());

    if !cursor.is_start_of_line()
        || !cursor.is_str("...")?
        || !matches!(cursor.peek_nth(3)?, None | Some('\r' | '\n' | '\t' | ' '))
    {
        return Ok(cursor.empty_span());
    }

    cursor.eat_str("...")?;
    let span = cursor.token();
    receiver.token(Token::DocumentEnd, span);

    trivia::trailing_lines(cursor, receiver)?;
    Ok(span)
}

fn directives<'s>(
    cursor: &mut Cursor<'s>,
    receiver: &mut (impl Receiver + ?Sized),
    document: &mut Document<'s>,
) -> Result<(), Diagnostic> {
    while cursor.is_char(char::DIRECTIVE)? {
        if !cursor.is_start_of_line() {
            receiver.diagnostic(Diagnostic::new(
                DiagnosticKind::DirectiveNotAtStartOfLine,
                cursor.next_span(),
            ));
        }

        token_char(cursor, receiver, Token::Directive, char::DIRECTIVE)?;

        if let Err(diag) = directive(cursor, receiver, document) {
            recover(cursor, receiver, diag, |cursor| {
                Ok(cursor.is(char::r#break)?
                    || (cursor.is_separated() && cursor.is_char(char::COMMENT)?))
            })?;
        }

        trivia::trailing_lines(cursor, receiver)?;
    }

    Ok(())
}

fn directive<'s>(
    cursor: &mut Cursor<'s>,
    receiver: &mut (impl Receiver + ?Sized),
    document: &mut Document<'s>,
) -> Result<(), Diagnostic> {
    debug_assert!(cursor.is_token_boundary());
    if cursor.is(char::non_space)? {
        let mut name = CowBuilder::new(cursor);
        name.push_while(cursor, char::non_space)?;
        let name = name.build();
        let name_span = cursor.token();
        receiver.token(Token::DirectiveName, name_span);

        match name.as_ref() {
            "YAML" => yaml_directive(cursor, receiver, document, name_span)?,
            "TAG" => tag_directive(cursor, receiver, document)?,
            _ => reserved_directive(cursor, receiver, name, name_span)?,
        }

        Ok(())
    } else {
        Err(Diagnostic::expected_token(Token::DirectiveName, cursor))
    }
}

fn yaml_directive<'s>(
    cursor: &mut Cursor<'s>,
    receiver: &mut (impl Receiver + ?Sized),
    document: &mut Document<'s>,
    name_span: Span,
) -> Result<(), Diagnostic> {
    if !trivia::try_separate_in_line(cursor, receiver)? {
        return Err(Diagnostic::expected_token(Token::YamlVersion, cursor));
    }

    let mut version = CowBuilder::new(cursor);
    let (major, minor) = yaml_version(cursor, &mut version)?;
    let version = version.build();
    let version_span = cursor.token();
    receiver.token(Token::YamlVersion, version_span);

    if document.version.is_some() {
        receiver.diagnostic(Diagnostic::new(
            DiagnosticKind::DuplicateYamlDirective,
            cursor.span(name_span.start),
        ));
    } else {
        document.version = Some(version);
    }

    match (major, minor) {
        (1, 1 | 2) => (),
        (1, _) => receiver.diagnostic(Diagnostic::new(
            DiagnosticKind::UnknownMinorVersion,
            version_span,
        )),
        (_, _) => receiver.diagnostic(Diagnostic::new(
            DiagnosticKind::UnknownMajorVersion,
            version_span,
        )),
    }

    while let Some(param_span) = try_directive_param(cursor, receiver)? {
        receiver.diagnostic(Diagnostic::new(
            DiagnosticKind::UnexpectedDiagnosticParameter,
            param_span,
        ))
    }

    Ok(())
}

fn yaml_version<'s>(
    cursor: &mut Cursor<'s>,
    value: &mut CowBuilder<'s>,
) -> Result<(u32, u32), Diagnostic> {
    let minor = yaml_version_part(cursor, value)?;

    if !cursor.is_char('.')? {
        return Err(Diagnostic::expected(Expected::Char('.'), cursor));
    }
    value.push(cursor);

    let major = yaml_version_part(cursor, value)?;

    if cursor.is(char::non_space)? {
        return Err(Diagnostic::expected(Expected::DecimalDigit, cursor));
    }

    Ok((minor, major))
}

fn yaml_version_part<'s>(
    cursor: &mut Cursor<'s>,
    value: &mut CowBuilder<'s>,
) -> Result<u32, Diagnostic> {
    let start = cursor.location();
    let value_start = value.len();
    if !cursor.is(char::dec_digit)? {
        return Err(Diagnostic::expected(Expected::DecimalDigit, cursor));
    }
    value.push_while(cursor, char::dec_digit)?;

    value.as_ref()[value_start..].parse::<u32>().map_err(|_| {
        Diagnostic::new(
            DiagnosticKind::VersionOverflow(value.as_ref().to_owned().into()),
            cursor.span(start),
        )
    })
}

fn tag_directive<'s>(
    cursor: &mut Cursor<'s>,
    receiver: &mut (impl Receiver + ?Sized),
    document: &mut Document<'s>,
) -> Result<(), Diagnostic> {
    trivia::separate_in_line(cursor, receiver)?;

    let handle_start = cursor.location();
    let handle = tag::handle(cursor, receiver)?;
    let handle_span = cursor.span(handle_start);

    trivia::separate_in_line(cursor, receiver)?;
    let prefix = tag::prefix(cursor, receiver)?;

    match document.tags.entry(handle) {
        btree_map::Entry::Vacant(entry) => {
            entry.insert(prefix);
        }
        btree_map::Entry::Occupied(entry) => {
            receiver.diagnostic(Diagnostic::new(
                DiagnosticKind::DuplicateTagDirective(entry.key().as_ref().to_owned().into()),
                handle_span,
            ));
        }
    }

    Ok(())
}

fn reserved_directive<'s>(
    cursor: &mut Cursor<'s>,
    receiver: &mut (impl Receiver + ?Sized),
    name: Cow<'s, str>,
    name_span: Span,
) -> Result<(), Diagnostic> {
    receiver.diagnostic(Diagnostic::new(
        DiagnosticKind::UnknownDirective(name.into_owned().into()),
        name_span,
    ));

    while try_directive_param(cursor, receiver)?.is_some() {}

    Ok(())
}

fn try_directive_param(
    cursor: &mut Cursor<'_>,
    receiver: &mut (impl Receiver + ?Sized),
) -> Result<Option<Span>, Diagnostic> {
    if trivia::try_separate_in_line(cursor, receiver)?
        && cursor.is(char::non_space)?
        && !cursor.is_char(char::COMMENT)?
    {
        cursor.eat_while(char::non_space)?;
        let span = cursor.token();
        receiver.token(Token::DirectiveParameter, span);
        Ok(Some(span))
    } else {
        Ok(None)
    }
}

impl<'s> Document<'s> {
    pub fn explicit(&self) -> bool {
        self.explicit
    }

    pub fn version(&self) -> Option<&Cow<'s, str>> {
        self.version.as_ref()
    }
}
