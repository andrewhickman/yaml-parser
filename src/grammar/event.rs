use alloc::vec::Vec;

use crate::{
    cursor::Cursor,
    grammar::{document, State},
    parser::Buffer,
    Diagnostic, Event, Receiver, Span,
};

pub(crate) fn event<'s>(
    cursor: &mut Cursor<'s>,
    receiver: &mut (impl Receiver + ?Sized),
    buffer: &mut Buffer<'s>,
    states: &mut Vec<State<'s>>,
) -> Option<Result<(Event<'s>, Span), Diagnostic>> {
    let res = match states.pop()? {
        State::Error(err) => Err(err),
        State::Stream => stream(cursor, states),
        State::Document { prev_terminated } => document(cursor, receiver, states, prev_terminated),
        State::DocumentValue { .. } => todo!(),
        State::DocumentEnd => todo!(),
        State::BlockSequence { .. } => todo!(),
        State::FlowSequence { .. } => todo!(),
        State::BlockMapping { .. } => todo!(),
        State::FlowMapping { .. } => todo!(),
        State::BlockMappingValue { .. } => todo!(),
        State::FlowMappingValue { .. } => todo!(),
        State::FlowPair { .. } => todo!(),
        State::FlowPairEnd { .. } => todo!(),
    };

    if res.is_err() {
        // todo recovery
    }

    Some(res)
}

fn stream<'s>(
    cursor: &mut Cursor<'s>,
    states: &mut Vec<State<'s>>,
) -> Result<(Event<'s>, Span), Diagnostic> {
    states.push(State::Document {
        prev_terminated: true,
    });

    Ok((
        Event::StreamStart {
            encoding: cursor.encoding(),
        },
        cursor.empty_span(),
    ))
}

fn document<'s>(
    cursor: &mut Cursor<'s>,
    receiver: &mut (impl Receiver + ?Sized),
    states: &mut Vec<State<'s>>,
    prev_terminated: bool,
) -> Result<(Event<'s>, Span), Diagnostic> {
    while !cursor.is_end_of_input()? {
        let (document, span) = document::prefix(cursor, receiver, prev_terminated)?;
        if document.explicit() || document::suffix(cursor, receiver)?.is_empty() {
            let version = document.version().cloned();
            states.push(State::DocumentValue { document });
            return Ok((Event::DocumentStart { version }, span));
        }
    }

    Ok((Event::StreamEnd, cursor.empty_span()))
}

fn document_value<'s>(
    cursor: &mut Cursor<'s>,
    receiver: &mut (impl Receiver + ?Sized),
    buffer: &mut Buffer<'s>,
    states: &mut Vec<State<'s>>,
    mut prev_terminated: bool,
) -> Result<(Event<'s>, Span), Diagnostic> {
    while !cursor.is_end_of_input()? {
        let (document, span) = document::prefix(cursor, receiver, prev_terminated)?;
        if document.explicit() || document::suffix(cursor, receiver)?.is_empty() {
            let version = document.version().cloned();
            states.push(State::DocumentValue { document });
            return Ok((Event::DocumentStart { version }, span));
        }
    }

    Ok((Event::StreamEnd, cursor.empty_span()))
}
