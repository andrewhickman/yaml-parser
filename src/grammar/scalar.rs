
#[derive(Debug, Copy, Clone, Hash, PartialEq, Eq)]
enum Chomping {
    Strip,
    Clip,
    Keep,
}