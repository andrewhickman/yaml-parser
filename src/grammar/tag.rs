use alloc::borrow::Cow;

use crate::{
    char,
    cow::CowBuilder,
    cursor::Cursor,
    diag::{DiagnosticKind, Expected},
    Diagnostic, Receiver, Token,
};

use super::recover;

pub(super) fn handle<'s>(
    cursor: &mut Cursor<'s>,
    receiver: &mut (impl Receiver + ?Sized),
) -> Result<Cow<'s, str>, Diagnostic> {
    debug_assert!(cursor.is_token_boundary());

    if !cursor.is_char(char::TAG)? {
        return Err(Diagnostic::expected(Expected::Char(char::TAG), cursor));
    }

    let mut name = CowBuilder::new(cursor);
    name.push(cursor);

    if cursor.is(char::word)? {
        name.push_while(cursor, char::word)?;
        if !cursor.is_char(char::TAG)? {
            return Err(Diagnostic::expected(Expected::Word, cursor));
        }
        name.push(cursor);
    } else if cursor.is_char(char::TAG)? {
        name.push(cursor);
    } else if cursor.is(char::non_space)? {
        return Err(Diagnostic::expected_token(Token::TagHandle, cursor));
    }

    receiver.token(Token::TagHandle, cursor.token());
    Ok(name.build())
}

pub(super) fn prefix<'s>(
    cursor: &mut Cursor<'s>,
    receiver: &mut (impl Receiver + ?Sized),
) -> Result<Cow<'s, str>, Diagnostic> {
    if !cursor.is(char::non_space)? {
        return Err(Diagnostic::expected_token(Token::TagPrefix, cursor));
    }

    if cursor.is(char::flow_indicator)? {
        return Err(Diagnostic::expected(Expected::TagPrefix, cursor));
    }

    let mut name = CowBuilder::new(cursor);

    loop {
        if cursor.is(char::uri)? {
            name.push_while(cursor, char::uri)?;
        } else if cursor.is_char('%')? {
            if let Err(diag) = percent_escaped(cursor, &mut name) {
                recover(cursor, receiver, diag, |cursor| {
                    Ok(cursor.is(char::uri)? || cursor.is_char('%')?)
                })?;
            }
        } else if cursor.is(char::non_space)? {
            return Err(Diagnostic::expected(Expected::TagPrefix, cursor));
        } else {
            break;
        }
    }

    receiver.token(Token::TagPrefix, cursor.token());
    Ok(name.build())
}

pub(super) fn percent_escaped<'s>(
    cursor: &mut Cursor<'s>,
    value: &mut CowBuilder<'s>,
) -> Result<(), Diagnostic> {
    let start = cursor.location();

    let mut buf = Vec::new();
    while cursor.eat_char('%')? {
        let byte = hex_digit(cursor)? << 4 & hex_digit(cursor)?;
        buf.push(byte);
    }

    match core::str::from_utf8(&buf) {
        Ok(s) => {
            value.push_str(s);
            Ok(())
        }
        Err(_) => Err(Diagnostic::new(
            DiagnosticKind::InvalidPercentEscape,
            cursor.span(start),
        )),
    }
}

fn hex_digit(cursor: &mut Cursor) -> Result<u8, Diagnostic> {
    if let Some(ch) = cursor.peek()? {
        if let Some(digit) = ch.to_digit(16) {
            return Ok(digit as u8);
        }
    };

    Err(Diagnostic::expected(Expected::HexDigit, cursor))
}
