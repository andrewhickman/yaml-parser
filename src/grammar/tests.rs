use serde::Serialize;

use crate::{cursor::Cursor, stream::Stream, Diagnostic, Location, Receiver, Span, Token};

#[derive(Serialize)]
#[serde(transparent)]
struct TestReceiver<T> {
    items: Vec<TestItem<T>>,
}

#[derive(Serialize)]
#[serde(untagged)]
enum TestItem<T> {
    Diagnostic { diag: Diagnostic },
    Token { token: Token, span: Span },
    Value { value: T },
}

impl<T> Receiver for TestReceiver<T> {
    fn diagnostic(&mut self, diag: Diagnostic) {
        self.items.push(TestItem::Diagnostic { diag })
    }

    fn token(&mut self, token: Token, span: Span) {
        self.items.push(TestItem::Token { token, span })
    }
}

impl<T> TestReceiver<T> {
    fn new() -> Self {
        TestReceiver { items: Vec::new() }
    }

    fn finish(&mut self, result: Result<T, Diagnostic>) {
        match result {
            Ok(value) => self.items.push(TestItem::Value { value }),
            Err(diag) => self.items.push(TestItem::Diagnostic { diag }),
        }
    }
}

pub(super) fn parse<'s, T, F>(f: F, s: &'s str) -> impl Serialize + 's
where
    T: Serialize + 's,
    F: Fn(&mut Cursor<'s>, &mut (dyn Receiver + 's)) -> Result<T, Diagnostic>,
{
    let mut receiver = TestReceiver::<T>::new();
    let mut cursor = Cursor::new(Stream::from_str(s));

    let res = f(&mut cursor, &mut receiver);
    receiver.finish(res);

    let mut location = Location::default();
    let mut prev_token = None;
    for item in &receiver.items {
        if let &TestItem::Token { span, token } = item {
            assert_eq!(location, span.start);
            assert!(prev_token != Some(token) || token == Token::Error);
            location = span.end;
            prev_token = Some(token);
        }
    }
    assert_eq!(location, cursor.location());
    assert!(cursor.is_end_of_input().unwrap());

    receiver
}
