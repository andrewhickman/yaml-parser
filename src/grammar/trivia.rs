#[cfg(test)]
mod tests;

use crate::{char, cursor::Cursor, diag::Expected, Diagnostic, Receiver, Span, Token};

use super::recover;

pub(super) fn try_non_content_break(
    cursor: &mut Cursor,
    receiver: &mut (impl Receiver + ?Sized),
) -> Result<bool, Diagnostic> {
    if try_break(cursor)? {
        receiver.token(Token::Break, cursor.token());
        Ok(true)
    } else {
        Ok(false)
    }
}

pub(super) fn non_content_break(
    cursor: &mut Cursor,
    receiver: &mut (impl Receiver + ?Sized),
) -> Result<(), Diagnostic> {
    if try_non_content_break(cursor, receiver)? {
        Ok(())
    } else {
        Err(Diagnostic::expected_token(Token::Break, cursor))
    }
}

fn try_break(cursor: &mut Cursor) -> Result<bool, Diagnostic> {
    if cursor.eat_char('\r')? {
        cursor.eat_char('\n')?;
        Ok(true)
    } else if cursor.eat_char('\n')? {
        Ok(true)
    } else {
        Ok(false)
    }
}

pub(super) fn separator_lines(
    cursor: &mut Cursor,
    receiver: &mut (impl Receiver + ?Sized),
) -> Result<(), Diagnostic> {
    if try_separate_in_line(cursor, receiver)? && cursor.is_char(char::COMMENT)? {
        comment_text(cursor, receiver)?;
    }

    if !cursor.is_end_of_input()? && !try_non_content_break(cursor, receiver)? && !cursor.is_separated() {
        return Err(Diagnostic::expected_token(Token::Break, cursor));
    }

    comment_lines(cursor, receiver)
}

pub(super) fn comment_lines(
    cursor: &mut Cursor,
    receiver: &mut (impl Receiver + ?Sized),
) -> Result<(), Diagnostic> {
    while try_separate_in_line(cursor, receiver)? {
        if cursor.is_char(char::COMMENT)? {
            comment_text(cursor, receiver)?;
        } else if !try_non_content_break(cursor, receiver)? || cursor.is_end_of_input()? {
            break;
        }
    }

    Ok(())
}

pub(super) fn comment_text(
    cursor: &mut Cursor,
    receiver: &mut (impl Receiver + ?Sized),
) -> Result<(), Diagnostic> {
    if !cursor.eat_char(char::COMMENT)? {
        return Err(Diagnostic::expected_token(Token::Comment, cursor));
    }
    cursor.eat_while(char::non_break)?;

    receiver.token(Token::Comment, cursor.token());

    if try_non_content_break(cursor, receiver)? || cursor.is_end_of_input()? {
        Ok(())
    } else {
        recover(
            cursor,
            receiver,
            Diagnostic::expected(Expected::Printable, cursor),
            |cursor| cursor.is(char::r#break),
        )?;
        try_non_content_break(cursor, receiver)?;
        Ok(())
    }
}

pub(super) fn separate_in_line(
    cursor: &mut Cursor,
    receiver: &mut (impl Receiver + ?Sized),
) -> Result<(), Diagnostic> {
    if try_separate_in_line(cursor, receiver)? {
        Ok(())
    } else {
        Err(Diagnostic::expected_token(Token::Separator, cursor))
    }
}

pub(super) fn try_separate_in_line(
    cursor: &mut Cursor,
    receiver: &mut (impl Receiver + ?Sized),
) -> Result<bool, Diagnostic> {
    if cursor.is(char::space)? {
        while cursor.is(char::space)? {
            cursor.bump();
        }
        receiver.token(Token::Separator, cursor.token());
        Ok(true)
    } else {
        Ok(cursor.is_separated())
    }
}
