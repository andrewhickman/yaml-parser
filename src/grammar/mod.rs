mod document;
mod event;
mod scalar;
mod tag;
#[cfg(test)]
mod tests;
mod trivia;

use crate::{cursor::Cursor, Diagnostic, Receiver, Span, Token};

use self::document::Document;
pub(crate) use self::event::event;

#[derive(Debug, Clone)]
pub(crate) enum State<'s> {
    Error(Diagnostic),
    Stream,
    Document {
        prev_terminated: bool,
    },
    DocumentValue {
        document: Document<'s>,
    },
    DocumentEnd,
    BlockSequence {
        indent: i32,
        context: Context,
        first: bool,
    },
    FlowSequence {
        indent: i32,
        context: Context,
        first: bool,
    },
    BlockMapping {
        indent: i32,
        context: Context,
        first: bool,
    },
    FlowMapping {
        indent: i32,
        context: Context,
        first: bool,
    },
    BlockMappingValue {
        explicit: bool,
        indent: i32,
        context: Context,
    },
    FlowMappingValue {
        allow_adjacent: bool,
        allow_empty: bool,
        indent: i32,
        context: Context,
    },
    FlowPair {
        indent: i32,
        context: Context,
    },
    FlowPairEnd {
        indent: i32,
        context: Context,
    },
}

#[derive(Debug, Copy, Clone, Hash, PartialEq, Eq)]
pub(crate) enum Context {
    BlockIn,
    BlockOut,
    BlockKey,
    FlowIn,
    FlowOut,
    FlowKey,
}

fn try_token_char(
    cursor: &mut Cursor,
    receiver: &mut (impl Receiver + ?Sized),
    token: Token,
    ch: char,
) -> Result<bool, Diagnostic> {
    debug_assert!(cursor.is_token_boundary());
    if cursor.eat_char(ch)? {
        receiver.token(token, cursor.token());
        Ok(true)
    } else {
        Ok(false)
    }
}

fn token_char(
    cursor: &mut Cursor,
    receiver: &mut (impl Receiver + ?Sized),
    token: Token,
    ch: char,
) -> Result<(), Diagnostic> {
    debug_assert!(cursor.is_token_boundary());
    if try_token_char(cursor, receiver, token, ch)? {
        Ok(())
    } else {
        Err(Diagnostic::expected_token(token, cursor))
    }
}

fn token(
    cursor: &mut Cursor,
    receiver: &mut (impl Receiver + ?Sized),
    token: Token,
    pred: impl Fn(char) -> bool + Clone,
) -> Result<Span, Diagnostic> {
    debug_assert!(cursor.is_token_boundary());
    while cursor.eat(pred.clone())? {}
    let span = cursor.token();
    receiver.token(token, span);
    Ok(span)
}

fn lookahead(
    cursor: &mut Cursor,
    pred: impl Fn(&mut Cursor) -> Result<bool, Diagnostic>,
) -> Result<bool, Diagnostic> {
    let mut start = cursor.clone();
    if pred(cursor)? {
        Ok(true)
    } else {
        *cursor = start;
        Ok(false)
    }
}

fn recover(
    cursor: &mut Cursor,
    receiver: &mut (impl Receiver + ?Sized),
    diag: Diagnostic,
    pred: impl Fn(&Cursor) -> Result<bool, Diagnostic>,
) -> Result<(), Diagnostic> {
    if diag.is_recoverable() {
        while !pred(cursor)? && !cursor.is_end_of_input()? {
            cursor.bump();
        }
        let span = cursor.token();
        if !span.is_empty() {
            receiver.token(Token::Error, span);
        }

        receiver.diagnostic(diag.clone());
        Ok(())
    } else {
        Err(diag)
    }
}
