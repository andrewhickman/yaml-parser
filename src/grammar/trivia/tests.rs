use insta::assert_yaml_snapshot;

use crate::grammar::{
    tests::{parse, parse_unseparated},
    trivia::{comment_lines, non_content_break, trailing_lines},
};

#[test]
fn test_line_break() {
    assert_yaml_snapshot!(parse(non_content_break, "\n"));
    assert_yaml_snapshot!(parse(non_content_break, "\r"));
    assert_yaml_snapshot!(parse(non_content_break, "\r\n"));
    assert_yaml_snapshot!(parse(non_content_break, "\0"));
    assert_yaml_snapshot!(parse(non_content_break, "foo"));
}

#[test]
fn test_comment_lines() {
    assert_yaml_snapshot!(parse(comment_lines, ""));
    assert_yaml_snapshot!(parse(comment_lines, "# comment"));
    assert_yaml_snapshot!(parse(comment_lines, "\n"));
    assert_yaml_snapshot!(parse(comment_lines, "\r\n"));
    assert_yaml_snapshot!(parse(comment_lines, "# comment\n"));
    assert_yaml_snapshot!(parse(comment_lines, " # comment"));
    assert_yaml_snapshot!(parse(comment_lines, "\t# comment"));
    assert_yaml_snapshot!(parse(comment_lines, " \n"));
    assert_yaml_snapshot!(parse(comment_lines, "\t\r\n\n\r\r\n\r\r"));
    assert_yaml_snapshot!(parse(comment_lines, "# comment\n"));
    assert_yaml_snapshot!(parse(comment_lines, "# comment\nfoo"));
    assert_yaml_snapshot!(parse(comment_lines, "foo"));
    assert_yaml_snapshot!(parse(comment_lines, "  # one\r\n# two\n\t#three\r   "));
    assert_yaml_snapshot!(parse(comment_lines, "# comment\n\n# two"));
    assert_yaml_snapshot!(parse(comment_lines, "# comment\0"));
    assert_yaml_snapshot!(parse(comment_lines, "# comment\0\r\n"));
}

#[test]
fn test_separator_lines() {
    assert_yaml_snapshot!(parse(trailing_lines, ""));
    assert_yaml_snapshot!(parse(trailing_lines, "#one"));
    assert_yaml_snapshot!(parse(trailing_lines, "#one\n"));
    assert_yaml_snapshot!(parse(trailing_lines, "#one\n#two"));
    assert_yaml_snapshot!(parse(trailing_lines, "foo"));
    assert_yaml_snapshot!(parse(trailing_lines, "\nfoo"));
    assert_yaml_snapshot!(parse_unseparated(trailing_lines, ""));
    assert_yaml_snapshot!(parse_unseparated(trailing_lines, "#one"));
    assert_yaml_snapshot!(parse_unseparated(trailing_lines, "#one\n"));
    assert_yaml_snapshot!(parse_unseparated(trailing_lines, "#one\n#two"));
    assert_yaml_snapshot!(parse_unseparated(trailing_lines, "foo"));
    assert_yaml_snapshot!(parse_unseparated(trailing_lines, "\nfoo"));
    assert_yaml_snapshot!(parse_unseparated(trailing_lines, " "));
    assert_yaml_snapshot!(parse_unseparated(trailing_lines, " #one"));
    assert_yaml_snapshot!(parse_unseparated(trailing_lines, " #one\n"));
    assert_yaml_snapshot!(parse_unseparated(trailing_lines, " #one\n#two"));
    assert_yaml_snapshot!(parse_unseparated(trailing_lines, " foo"));
    assert_yaml_snapshot!(parse_unseparated(trailing_lines, " \nfoo"));
    assert_yaml_snapshot!(parse_unseparated(trailing_lines, " #comment\nfoo"));
}
