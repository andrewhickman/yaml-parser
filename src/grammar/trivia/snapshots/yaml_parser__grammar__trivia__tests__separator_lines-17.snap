---
source: src/grammar/trivia/tests.rs
expression: "parse_unseparated(trailing_lines, \" foo\")"
---
- token: separator
  stream: " "
  span:
    start:
      index: 0
      line: 0
      column: 0
    end:
      index: 1
      line: 0
      column: 1
- value: ~
- remainder: foo
  span:
    start:
      index: 1
      line: 0
      column: 1
    end:
      index: 4
      line: 0
      column: 4

